File: src/Neighbor/Neighbor.cc
//---------------------------------Spheral++----------------------------------//
// Neighbor -- Abstract interface base class for the Neighbor objects.
//
// Created by J. Michael Owen, Sun Nov 12 10:33:55 2000
//----------------------------------------------------------------------------//
#include "Neighbor.hh"

#include "Geometry/Dimension.hh"
#include "Geometry/GeomPlane.hh"
#include "Utilities/DBC.hh"
#include "Field/Field.hh"
#include "Field/FieldList.hh"
#include "NodeList/NodeList.hh"
#include "NodeList/NodeListRegistrar.hh"
#include "Utilities/testBoxIntersection.hh"

#include <algorithm>
using std::vector;
using std::cout;
using std::cerr;
using std::endl;
using std::min;
using std::max;
using std::abs;

namespace Spheral {

//------------------------------------------------------------------------------
// Construct with the given NodeList and search type.
//------------------------------------------------------------------------------
template<typename Dimension>
Neighbor<Dimension>::
Neighbor(NodeList<Dimension>& nodeList,
         const NeighborSearchType searchType,
         const double kernelExtent):
  mSearchType(searchType),
  mKernelExtent(kernelExtent),
  mNodeListPtr(&nodeList),
  mNodeExtent("Node Extent", nodeList) {
  mNodeListPtr->registerNeighbor(*this);
}

//------------------------------------------------------------------------------
// Destructor.
//------------------------------------------------------------------------------
template<typename Dimension>
Neighbor<Dimension>::
~Neighbor() {
  if (mNodeListPtr != nullptr) mNodeListPtr->unregisterNeighbor();
}

//------------------------------------------------------------------------------
// Get the type of search.
//------------------------------------------------------------------------------
template<typename Dimension>
NeighborSearchType
Neighbor<Dimension>::
neighborSearchType() const {
  return mSearchType;
}

//------------------------------------------------------------------------------
// Set the type of search.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
neighborSearchType(NeighborSearchType searchType) {
  mSearchType = searchType;
}

//------------------------------------------------------------------------------
// Return the field of node extents.
//------------------------------------------------------------------------------
template<typename Dimension>
const Field<Dimension, typename Dimension::Vector>&
Neighbor<Dimension>::
nodeExtentField() const {
  return mNodeExtent;
}

// Allow read/write access to the node extent Field for descendent classes.
template<typename Dimension>
Field<Dimension, typename Dimension::Vector>&
Neighbor<Dimension>::
accessNodeExtentField() {
  return mNodeExtent;
}

//------------------------------------------------------------------------------
// Access the node list.
//------------------------------------------------------------------------------
template<typename Dimension>
const NodeList<Dimension>&
Neighbor<Dimension>::
nodeList() const {
  CHECK(mNodeListPtr != nullptr);
  return *mNodeListPtr;
}

template<typename Dimension>
const NodeList<Dimension>*
Neighbor<Dimension>::
nodeListPtr() const {
  return mNodeListPtr;
}

//------------------------------------------------------------------------------
// Set the node list.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
nodeList(NodeList<Dimension>& nodeList) {
  CHECK(&nodeList);
  mNodeListPtr = &nodeList;
  mNodeExtent.setNodeList(nodeList);
}

template<typename Dimension>
void
Neighbor<Dimension>::
nodeListPtr(NodeList<Dimension>* nodeListPtr) {
  CHECK(nodeListPtr);
  mNodeListPtr = nodeListPtr;
  mNodeExtent.setNodeList(*nodeListPtr);
}

//------------------------------------------------------------------------------
// Unregister the node list.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
unregisterNodeList() {
  mNodeListPtr = nullptr;
}

//------------------------------------------------------------------------------
// Determine the extent of the given node's smoothing tensor.
//------------------------------------------------------------------------------
template<typename Dimension>
typename Dimension::Vector
Neighbor<Dimension>::
nodeExtent(int nodeID) const {
  CHECK(nodeID >= 0 and nodeID < nodeList().numNodes());
  return HExtent(nodeList().Hfield()(nodeID), kernelExtent());
}

//------------------------------------------------------------------------------
// Force the node extent field to be computed.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
setNodeExtents() {
  for (int nodeID = 0; nodeID < nodeList().numNodes(); ++nodeID) {
    mNodeExtent(nodeID) = nodeExtent(nodeID);
  }
}

template<typename Dimension>
void
Neighbor<Dimension>::
setNodeExtents(const vector<int>& nodeIDs) {
  for (typename vector<int>::const_iterator nodeIDItr = nodeIDs.begin();
       nodeIDItr < nodeIDs.end();
       ++nodeIDItr) {
    CHECK(*nodeIDItr >= 0 and *nodeIDItr < nodeList().numNodes());
    mNodeExtent(*nodeIDItr) = nodeExtent(*nodeIDItr);
  }
}

template<typename Dimension>
void
Neighbor<Dimension>::
setInternalNodeExtents() {
  for (int nodeID = 0; nodeID < nodeList().numInternalNodes(); ++nodeID) {
    mNodeExtent(nodeID) = nodeExtent(nodeID);
  }
}

template<typename Dimension>
void
Neighbor<Dimension>::
setGhostNodeExtents() {
  for (int nodeID = nodeList().firstGhostNode(); nodeID < nodeList().numNodes(); ++nodeID) {
    mNodeExtent(nodeID) = nodeExtent(nodeID);
  }
}

//------------------------------------------------------------------------------
// Set the master list of nodes based on a particular node index.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
setMasterList(int nodeID,
              std::vector<int>& masterList,
              std::vector<int>& coarseNeighbors) const {
  CHECK(valid());
  CHECK(nodeID >= 0 and nodeID < nodeList().numInternalNodes());
  const auto& positions = nodeList().positions();
  const auto& Hfield = nodeList().Hfield();
  this->setMasterList(positions(nodeID), Hfield(nodeID), masterList, coarseNeighbors);
}

//------------------------------------------------------------------------------
// Set the refined list of potential neighbors based on a particular node index.
//------------------------------------------------------------------------------
template<typename Dimension>
void
Neighbor<Dimension>::
setRefineNeighborList(int nodeID,
                      const std::vector<int>& coarseNeighbors,
                      std::vector<int>& refineNeighbors) const {
  CHECK(valid());
  CHECK(nodeID >= 0 and nodeID < nodeList().numInternalNodes());
  const auto& positions = nodeList().positions();
  const auto& Hfield = nodeList().Hfield();
  this->setRefineNeighborList(positions(nodeID), Hfield(nodeID), coarseNeighbors, refineNeighbors);
}

//------------------------------------------------------------------------------
// General precull routine.  Select nodes based on a range of positions and
// extents.
//------------------------------------------------------------------------------
template<typename Dimension>
vector<int>
Neighbor<Dimension>::
precullList(const Vector& minMasterPosition, const Vector& maxMasterPosition,
            const Vector& minMasterExtent, const Vector& maxMasterExtent,
            const vector<int>& coarseList) const {
  const auto n = coarseList.size();

  // Empty vector to accumulate the result in.
  vector<int> cullList;

  // Get a reference to the node positions and node extent field.
  const auto& positions = nodeList().positions();
  const auto& nodeExtents = nodeExtentField();

  // What kind of preculling we're doing determines the applied test.
  if (neighborSearchType() == NeighborSearchType::GatherScatter) {
    
    // Gather-Scatter.
// #pragma omp parallel
    // {
    //   vector<int> cullList_private;
// #pragma omp for nowait
      for (auto k = 0; k < n; ++k) {
        const auto  j = coarseList[k];
        const auto& nodePosition = positions(j);
        const auto  minNodeExtent = nodePosition - nodeExtents(j);
        const auto  maxNodeExtent = nodePosition + nodeExtents(j);
        if (testPointInBox(nodePosition,
                           minMasterExtent,
                           maxMasterExtent) or
            testBoxIntersection(minMasterPosition,
                                maxMasterPosition,
                                minNodeExtent,
                                maxNodeExtent)) {
          cullList.push_back(j);
        }
      }

// #pragma omp critical
//       std::copy(cullList_private.begin(), cullList_private.end(), std::back_inserter(cullList));
//     }

  } else if (neighborSearchType() == NeighborSearchType::Gather) {

    // Gather.
    for (typename vector<int>::const_iterator coarseItr = coarseList.begin();
         coarseItr != coarseList.end();
         ++coarseItr) {
      const auto  j = *coarseItr;
      const auto& nodePosition = positions(j);
      const auto  gatherTest = testPointInBox(nodePosition,
                                              minMasterExtent,
                                              maxMasterExtent);
      if (gatherTest) cullList.push_back(j);
    }

  } else {

    // Scatter.
    CHECK(neighborSearchType() == NeighborSearchType::Scatter);
    for (typename vector<int>::const_iterator coarseItr = coarseList.begin();
         coarseItr != coarseList.end();
         ++coarseItr) {
      const auto  j = *coarseItr;
      const auto& nodePosition = positions(j);
      const auto  minNodeExtent = nodePosition - nodeExtents(j);
      const auto  maxNodeExtent = nodePosition + nodeExtents(j);
      const auto  scatterTest = testBoxIntersection(minMasterPosition,
                                                    maxMasterPosition,
                                                    minNodeExtent,
                                                    maxNodeExtent);
      if (scatterTest) cullList.push_back(j);
    }

  }

  ENSURE(cullList.size() <= coarseList.size());
  return cullList;
}

// //------------------------------------------------------------------------------
// // Cull the local (to this NodeList) neighbor info based on the current master
// // state.
// // *NOTE* -- this is not safe to do when you want to use this neighbor info 
// // with different NodeLists!
// //------------------------------------------------------------------------------
// template<typename Dimension>
// void
// Neighbor<Dimension>::
// precullForLocalNodeList() {

//   // Grab the state.
//   const Field<Dimension, Vector>& r = this->nodeList().positions();
//   const Field<Dimension, Vector>& extent = nodeExtentField();

//   // Find the min/max master node positions and extents.
//   Vector minMasterPosition = DBL_MAX;
//   Vector maxMasterPosition = -DBL_MAX;
//   Vector minMasterExtent = DBL_MAX;
//   Vector maxMasterExtent = -DBL_MAX;
//   for (const_iterator masterItr = masterBegin();
//        masterItr != masterEnd();
//        ++masterItr) {
//     const int i = *masterItr;
//     const Vector& ri = r(i);
//     const Vector minExtent = ri - extent(i);
//     const Vector maxExtent = ri + extent(i);
//     for (int j = 0; j != Dimension::nDim; ++j) {
//       minMasterPosition(j) = min(minMasterPosition(j), ri(j));
//       maxMasterPosition(j) = max(maxMasterPosition(j), ri(j));
//       minMasterExtent(j) = min(minMasterExtent(j), minExtent(j));
//       maxMasterExtent(j) = max(maxMasterExtent(j), maxExtent(j));
//     }
//   }

//   // Now use this info to precull the coarse neighbor list.
//   *mCoarseNeighborListPtr = precullList(minMasterPosition, maxMasterPosition,
//                                         minMasterExtent, maxMasterExtent,
//                                         *mCoarseNeighborListPtr);
// }

//------------------------------------------------------------------------------
// Provide a basic test to determine whether the neighbor base is in a valid
// (i.e., ready to use) state.
//------------------------------------------------------------------------------
template<typename Dimension>
bool
Neighbor<Dimension>::
valid() const {
  return (kernelExtent() > 0.0);
}

// //------------------------------------------------------------------------------
// // Set the global bounding box.
// //------------------------------------------------------------------------------
// template<typename Dimension>
// void
// Neighbor<Dimension>::
// setBoundingBox() {
//   const NodeListRegistrar<Dimension>& registrar = NodeListRegistrar<Dimension>::instance();
//   FieldList<Dimension, Vector> positions(FieldListBase::Reference);
//   for (typename NodeListRegistrar<Dimension>::const_fluid_iterator itr = registrar.fluidBegin();
//        itr != registrar.fluidEnd();
//        ++itr) {
//     positions.appendField((**itr).positions());
//   }
//   globalBoundingBox(positions, mXmin, mXmax, false);
// }

}

File: src/Pybind11Wraps/Neighbor/Neighbor.py
#-------------------------------------------------------------------------------
# Neighbor
#-------------------------------------------------------------------------------
from PYB11Generator import *
from NeighborAbstractMethods import *

@PYB11template("Dimension")
class Neighbor:

    PYB11typedefs = """
    typedef typename %(Dimension)s::Scalar Scalar;
    typedef typename %(Dimension)s::Vector Vector;
    typedef typename %(Dimension)s::SymTensor SymTensor;
    typedef NodeList<%(Dimension)s> NodeListType;
    typedef GeomPlane<%(Dimension)s> Plane;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               nodeList = "NodeListType&",
               searchType = "const NeighborSearchType",
               kernelExtent = "const double"):
        "Construct a Neighbor with minimal info"

    #...........................................................................
    # Methods
    @PYB11returnpolicy("reference_internal")
    @PYB11const
    def nodeExtentField(self):
        "The internal field of node extents"
        return "const Field<%(Dimension)s, Vector>&"

    @PYB11const
    def nodeList(self):
        "The NodeList associated with this Neighbor"
        return "const NodeListType&"

    @PYB11pycppname("nodeList")
    def setnodeList(self, val="NodeListType&"):
        "Set the NodeList for this Neighbor"
        return "void"

    def unregisterNodeList(self):
        "Disassociate with the currently registered NodeList."
        return "void"

    @PYB11const
    def nodeExtent(self, nodeID="int"):
        "Return the node extent for the requested node ID in the associated NodeList"
        return "Vector"
                           
    def setNodeExtents(self):
        "Set the node extent field based on the current NodeList"
        return "void"

    @PYB11pycppname("setNodeExtents")
    def setNodeExtents1(self, nodeIDs="const std::vector<int>&"):
        "Set the node extent field for the given nodes in the current NodeList"
        return "void"

    def setInternalNodeExtents(self):
        "Set the node extent field for the internal nodes in the current NodeList"
        return "void"

    def setGhostNodeExtents(self):
        "Set the node extent field for the ghost nodes in the current NodeList"
        return "void"

    @PYB11pycppname("setMasterList")
    @PYB11virtual
    @PYB11const
    def setMasterList0(self,
                       nodeID = "int",
                       masterList = "std::vector<int>&",
                       coarseNeighbors = "std::vector<int>&",
                       ghostConnectivity = ("const bool", "false")):
        "Fill the given arrays with (master, coarse) neighbor info for the given node"
        return "void"

    @PYB11pycppname("setRefineNeighborList")
    @PYB11virtual
    @PYB11const
    def setRefineNeighborList0(self,
                               nodeID = "int",
                               coarseNeighbors = "const std::vector<int>&",
                               refineNeighbors = "std::vector<int>&"):
        "Fill the final array with refine neighbor info for the given node based on the coarse set"
        return "void"

    @PYB11const
    def precullList(self,
                    minMasterPosition = "const Vector&",
                    maxMasterPosition = "const Vector&",
                    minMasterExtent = "const Vector&",
                    maxMasterExtent = "const Vector&",
                    coarseNeighbors = "const std::vector<int>&"):
        "Return a culled list of potential neighbors based on the (min,max) info given"
        return "std::vector<int>"

    #...........................................................................
    # Virtual methods
    @PYB11virtual
    def reinitialize(self):
        "Reinitialize in case something changed."
        return "void"

    @PYB11virtual
    def reinitialize(self,
                     xmin = "const Vector&",
                     xmax = "const Vector&",
                     htarget = "const Scalar"):
        "Reinitialize to possibly more efficient based on the specified box (xmin,xmax) and htarget size"
        return "void"

    @PYB11virtual
    @PYB11const
    def valid(self):
        "Test if the Neighbor is valid, i.e., ready to be queried for connectivity information."
        return "bool"

    #...........................................................................
    # Static methods
    @PYB11static
    @PYB11pycppname("HExtent")
    def HExtent0(self,
                 H = "const Scalar&",
                 kernelExtent = "const double"):
        "Return the maximum extent in each Cartesian direction for the given H"
        return "Vector"

    @PYB11static
    @PYB11pycppname("HExtent")
    def HExtent1(self,
                 H = "const SymTensor&",
                 kernelExtent = "const double"):
        "Return the maximum extent in each Cartesian direction for the given H"
        return "Vector"

    #...........................................................................
    # Protected methods
    @PYB11protected
    def accessNodeExtentField(self):
        "Read/write access the protected node extent field member"
        return "Field<%(Dimension)s, Vector>&"

    #...........................................................................
    # Properties
    neighborSearchType = PYB11property("NeighborSearchType", "neighborSearchType", "neighborSearchType", doc="The search algorithm for nodes interacting")
    kernelExtent = PYB11property("double", "kernelExtent", "kernelExtent", doc="The kernel extent for nodes interacting in eta space")

#-------------------------------------------------------------------------------
# Add the abstract interface
#-------------------------------------------------------------------------------
PYB11inject(NeighborAbstractMethods, Neighbor, pure_virtual=True)
